ash/src/lib.rs

```diff
@@ -66,6 +66,40 @@ impl<'r, T> RawPtr<T> for Option<&'r T> {
     }
 }
 
+/// Given a mutable raw pointer to a type with an `s_type` member such as [`vk::BaseOutStructure`],
+/// match on a set of Vulkan structures. The struct will be rebound to the
+/// given variable of the type of the given Vulkan structure.
+///
+/// ```
+/// let mut info = ash::vk::DeviceCreateInfo::default();
+/// let info: *mut ash::vk::BaseOutStructure = <*mut _>::cast(&mut info);
+/// unsafe {
+///     ash::match_struct!(match info {
+///         info @ ash::vk::DeviceQueueCreateInfo => {
+///             dbg!(&info); // Unreachable
+///         }
+///         info @ ash::vk::DeviceCreateInfo => {
+///             dbg!(&info);
+///         }
+///     })
+/// }
+/// ```
+#[macro_export]
+macro_rules! match_struct {
+    (match $p:ident { $($bind:ident @ $ty:path => $body:block $(,)?)+ }) => {
```

Comment:
`$body:block` should probably be changed to `tt` (or is there a better type) to allow bodies without `{ .. }` block.

Reply:
I've been able to get this "working" for both blocks and expressions with either `stmt` or `expr`, but the downside is that it doesn't enforce a trailing comma and successfully parses something like:

```rust
            info @ ash::vk::ApplicationInfo => panic!()
            info @ ash::vk::InstanceCreateInfo => panic!()
```

(note that `tt` doesn't understand macro calls like `panic!()` when not enclosed in anything that resembles a block or expression.)

On the other hand, with the addition of `$(_ => $any:xxx $(,)?)?` for an optional default value the compiler correctly demands the use of an explicit comma:

```rust
error: `$body:stmt` may be followed by `_`, which is not allowed for `stmt` fragments
   --> ash/src/lib.rs:115:72
    |
115 |     (match $p:ident { $($bind:ident @ $ty:path => $body:stmt $(,)?)+ $(_ => $any:stmt $(,)?)? }) => {
    |                                                                        ^ not allowed after `stmt` fragments
    |
    = note: allowed there are: `=>`, `,` or `;`
```

Ideally we should have something that either parses an optional comma for `$body:block $(,)?` _or_ an expression which is forcibly terminated by a comma: `$body:expr,`. While we could have multiple `macro_rules!` arms for either method, I'm not sure how to write that in such a way that it can be alternated for any of the clauses a user puts in the macro.

Reply:
I think it's okay to support a reasonable subset of conventional syntax like this. Perhaps a more complete solution exists, but this is good enough and making declarative macros perfect can be very painful.

Reply:
Fortunately I had already updated the documentation to state that it _must_ be wrapped in curly braces and something like `_ => None` is currently not supported. Perhaps that triggers someone more fluent in declarative macros to PR a proper implementation :+1: